libs/libarchfpga/src/arch_util.h

```diff
@@ -89,4 +89,107 @@ bool block_type_contains_blif_model(t_logical_block_type_ptr type, const std::st
 //Returns true of a pb_type (or it's children) contain the specified blif model name
 bool pb_type_contains_blif_model(const t_pb_type* pb_type, const std::string& blif_model_name);
 
+// Physical tile and logical block types utility functions.
+
+bool is_opin(int ipin, t_physical_tile_type_ptr type);
+bool is_input_type(t_physical_tile_type_ptr type);
+bool is_output_type(t_physical_tile_type_ptr type);
+bool is_io_type(t_physical_tile_type_ptr type);
+
+//Returns the corresponding physical pin based on the input parameters:
+// - physical_tile
+// - relative_pin: this is the pin relative to a specific sub tile
+// - capacity location: absolute sub tile location
+int get_physical_pin_from_capacity_location(t_physical_tile_type_ptr physical_tile, int relative_pin, int capacity_location);
+
+//Returns a pair consisting of the absolute capacity location relative to the pin parameter
+//and the relative pin within the sub tile
+std::pair<int, int> get_capacity_location_from_physical_pin(t_physical_tile_type_ptr physical_tile, int pin);
+
+//Returns the name of the pin_index'th pin on the specified block type
+std::string block_type_pin_index_to_name(t_physical_tile_type_ptr type, int pin_index);
+
+//Returns the name of the class_index'th pin class on the specified block type
+std::vector<std::string> block_type_class_index_to_pin_names(t_physical_tile_type_ptr type, int class_index);
+
+//Returns the physical tile type matching a given physical tile type name, or nullptr (if not found)
+t_physical_tile_type_ptr find_tile_type_by_name(std::string name, const std::vector<t_physical_tile_type>& types);
+
+int find_pin_class(t_physical_tile_type_ptr type, std::string port_name, int pin_index_in_port, e_pin_type pin_type);
+
+int find_pin(t_physical_tile_type_ptr type, std::string port_name, int pin_index_in_port);
+
+//Returns the maximum number of pins within a logical block
+int get_max_num_pins(t_logical_block_type_ptr logical_block);
+
+//Verifies whether a given logical block is compatible with a given physical tile
+bool is_tile_compatible(t_physical_tile_type_ptr physical_tile, t_logical_block_type_ptr logical_block);
+
+//Verifies whether a logical block and a relative placement location is compatible with a given physical tile
+bool is_sub_tile_compatible(t_physical_tile_type_ptr physical_tile, t_logical_block_type_ptr logical_block, int sub_tile_loc);
+
+//Returns the physical tile type which 'best' matches logical_block
+t_physical_tile_type_ptr pick_best_physical_type(t_logical_block_type_ptr logical_block);
+
+//Returns the logical block type which 'best' matches the physical tile
+t_logical_block_type_ptr pick_best_logical_type(t_physical_tile_type_ptr physical_tile);
```

Comment:
Same comment as the function `pick_best_physical_type()`

Reply:
Done (see comment above)